### ISSUE TITLE
Expo AV loadAsync never rejects if a file fails to load

### ISSUE THREAD

**Issue by ghost** [expo/expo#20333]
### Summary

Howdy!  I am having the same issue described in #18627.  If Audio.Sound's loadAsync method is given a uri for a file which does not exist or can't be played, it never rejects.

I'm only working in iOS, so I'm not sure if this is true for Android or web.  I'm seeing it in both sdk version 46 and 47.

I tried the snack in the linked issue, and the reject does not occur on my simulator or when I run through an iOS device in the snack using the Tap to Play.  Real devices also have this issue.

### What platform(s) does this occur on?

iOS

### Environment
```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.6.1
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.18.0 - ~/.nvm/versions/node/v16.18.0/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v16.18.0/bin/yarn
      npm: 8.19.2 - ~/.nvm/versions/node/v16.18.0/bin/npm
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2021.3 AI-213.7172.25.2113.9123335
      Xcode: 14.1/14B47b - /usr/bin/xcodebuild
    npmPackages:
      expo: ~47.0.8 => 47.0.8
      react: 18.1.0 => 18.1.0
      react-native: 0.70.5 => 0.70.5
    npmGlobalPackages:
      expo-cli: 6.0.8
    Expo Workflow: managed
```
### Minimal reproducible example

Here's a delicious snack

https://snack.expo.dev/@wavy-jeff/load-ye-a-sound

For another example, here's a public repo

https://github.com/onwavy/sound-loading-example

**Comment by RRaideRR** [comment]
I'm having a similar issue on my expo-av Video component which I cannot track down. Sometimes, no alert pops up. Like every third or fourth try!

```
      videoRef.current
        .loadAsync({ uri: cachedVideo.localUri }, { isLooping: true })
        .then((e) => {
          Alert.alert('Not always printed');
        })
        .catch((e) => {
          Alert.alert('Error', JSON.stringify(e));
        });
```

**Comment by artsnr1** [comment]
It's still there in sdk 49...sigh!! Also if internet connectivity drops it keeps trying to load and never throws an error

**Comment by tgeorgiev** [comment]
experiencing the same with `createAsync` and a corrupted audio file. I have the file locally in documentDirectory, I can getInfoAsync and validate it exists, but when I call 
```
console.log('pre create');
await Audio.Sound.createAsync({ uri: uri });`
console.log('post create');
```

the whole execution stalls and in the console you can see just `pre create`.

Any hints to check if a file can be played, before calling createAsync, are appretiated!

**Comment by jefflieb-wavy** [comment]
Hello again bot.  This issue is still present in sdk 49 with the latest expo-av version.

**Comment by muratak18** [comment]
I'm having the same issue. 
Expo version 49.
On Android it works as expected. But on IOS it does not throw error.

**Comment by squareborg** [comment]
Same issue for me,  

Users on ios getting app hang as the promise never rejects. 

I think in my case the uri points to a file that iOS has made unavailable either through icloud removing the local file or some other feature in ios.

Anyone got any workarounds? Gonna have to see if maybe I can stat the file before loading it

**Comment by francescovenica** [comment]
what's the file name? I just found out if the file name has spaces it doesn't work...removing the spacing it started to work, maybe is something related to the encoding of the url?

**Comment by squareborg** [comment]
> what's the file name? I just found out if the file name has spaces it doesn't work...removing the spacing it started to work, maybe is something related to the encoding of the url?

Yeah you are getting a succesful load as you fixed your uri, but it should reject if the uri is invalid right, the problem is the promise never resolves or rejects just hangs. at least thats what I see, if you hardcode your uri to some incorrect location your promise reslove reject or hang?

**Comment by francescovenica** [comment]
yeah you are right, it should still throw the error or at least timeout is the url is wrong...I wrote just to give a workaround to who read this post (I've just spent the day on this 😂 )

**Comment by squareborg** [comment]
Im getting around it with RNFS for now. I'd imagine RNFS.exists would be enough but im not to sure how iOS manages storage if its in icloud,  so I went nukelear and made sure I could read the file, was a pretty horrible bug stopping the app launching

```
try {
    await RNFS.readFile(settings.customMp3Object.fileCopyUri, 'base64');
    await holdSound.current.loadAsync(
        { uri: settings.customMp3Object.fileCopyUri },
        {
            isLooping: true,
            volume: settings.holdMusicVolume
        }
    );
} catch (error) {
```

**Comment by squareborg** [comment]
> what's the file name? I just found out if the file name has spaces it doesn't work...removing the spacing it started to work, maybe is something related to the encoding of the url? replacing the string with `encodeURI(audioUrl)` fix all my issues

I just found out I have the same file name space bug too, Thanks i'll implement your suggestion, Android works fine with spaces

**Comment by squareborg** [comment]
> > what's the file name? I just found out if the file name has spaces it doesn't work...removing the spacing it started to work, maybe is something related to the encoding of the url? replacing the string with `encodeURI(audioUrl)` fix all my issues
> 
> I just checked, and found out I have the same file name space bug too, Thanks i'll implement your suggestion, Android works fine with spaces

Okay so of my space in the filename bug, I actuallly needed to use `decodeURI` for my RNFS (workaround) call and as my URI is alredy encoded by `react-native-document-picker` I didnt need to encode the uri for the expo-av .loadAsync() call

Seems RNFS doesnt want encoded URI's

**Comment by qusaieilouti99** [comment]
I can confirm that the issue still exists on SDK 49. It's an unacceptable major bug. Why has it been ignored for over a year? Has anyone found a workaround?

**Comment by squareborg** [comment]
> I can confirm that the issue still exists on SDK 49. It's an unacceptable major bug. Why has it been ignored for over a year? Has anyone found a workaround?

Far from ideal but I had to go with:

```
const res = await Promise.race([
    holdSound.current.loadAsync(
        { uri: filePath },
        {
            isLooping: true,
            volume: settings.holdMusicVolume
        }
    ),
    new Promise((resolve, _) =>
        setTimeout(() => {
            resolve('error');
        }, 5000)
    )
]);

if (res === 'error') {
    // handle timeout
}
```

**Comment by qusaieilouti99** [comment]
> > I can confirm that the issue still exists on SDK 49. It's an unacceptable major bug. Why has it been ignored for over a year? Has anyone found a workaround?
> 
> Far from ideal but I had to go with `Promise.race()`:
> 
> ```
> const res = await Promise.race([
>     holdSound.current.loadAsync(
>         { uri: filePath },
>         {
>             isLooping: true,
>             volume: settings.holdMusicVolume
>         }
>     ),
>     new Promise((resolve, reject) =>
>         setTimeout(() => {
>             reject(new Error("timed out"));
>         }, 5000)
>     )
> ]);
> ```

Thank you for your reply, but can't go with this because I'm loading a bunch of audio files simultaneously and this can lead to a lot of timers being added into memory at the same time, I'm in a performance enhancement cycle of my app and this is the last thing I could do

**Comment by jefflieb-wavy** [comment]
Preventing this from closing.  Updated the snack to use the latest av version, the issue continues clinging tenaciously to its existence.

**Comment by TrustyTechSG** [comment]
This is still exist with SDK 51.0.18, and the issue only happen on iOS (both simulator and real device). I can confirm my file exists with ExpoFileSystem.getInfoAsync return exists = true. but the same local file uri never return with loadAsync({ uri }) call. This issue only on iOS. Android works all fine.

Here is a example of my file uri:
file:///Users/xxxx/Library/Developer/CoreSimulator/Devices/XXX-XXX-XXX/data/Containers/Data/Application/XXX-XXX-XXX/Library/Caches/123456.ogg

**Comment by nikkwong** [comment]
I'm running into this issue as well. Is the team refusing to chime in?

**Comment by JCcastagne** [comment]
Still is an issue

**Comment by christophemenager** [comment]
I confirm this is still an issue in latest expo SDK 51 and react-native 0.75.3, I went for the promise workaround even if not ideal

**Comment by mfkrause** [comment]
We're also experiencing this on Expo SDK 52. We're preloading a bunch of sounds on app start and, intermittently, this causes the app to hang for either several seconds or indefinitely because the Promise never resolves or rejects.

For what it's worth, the new `expo-audio` library does not seem to have this issue. We've refactored our codebase to use this library (which is still in beta) and it seems to be stable. Our audio use case is pretty simple, so the new library works well, might not be the case for everyone, though.

**Comment by brunoccc** [comment]
Same here: often the library gets stuck inside loadAsync and never returns. 
Potentially I could try the `Promise.race` workaround, but I am worried about performances and also possible memory leaks (i.e. what actually happened to the "pending" loadAsync that never returned?).
I also tried to use the new `expo-audio` but stumbled into even bigger problems (i.e. the app crashes saying it can't find the native module). And the docs send me to a README file that doesn't exist.

**Comment by singlecolor** [comment]
Today we found a root cause for this issue in our app (SDK 52), and we have a workaround that at least is working for our use case.

We discovered that loadAsync(), at least on iOS 18+, has a **limit of 18 concurrently loaded sound files**. I am not certain if this is a specified limit anywhere, or differs across devices. At least on an iPhone XR on iOS 18.3.1 in dev builds and production builds, that is the case.

We have been loading all music tracks and sfx on app load for all of our apps to date, but the most we ever had in an app was just 13. Our current app has 26. We never use unloadAsync(), which is certainly lazy, but at one point we set our sounds up based on Expo's old documentation (SDK ~40 I'd guess). We never had this issue so never changed it.

We load our sounds via initSounds() when our App.tsx is loaded. I am showing only 2 sounds here, but we have 26.

```
export let soundsDict = {
  // Sound Effects
  click: { audio: new Audio.Sound(), isLoaded: false },
  back: { audio: new Audio.Sound(), isLoaded: false },
}
```

We load our sounds 

```
export function initSounds() {
  try {
    let soundOpts = {
      volume: 1,
      rate: 1
    }

    // Setup Audio - Sound Effects
    soundsDict['click'].audio.loadAsync(require('../assets/sounds/ButtonPress_Simple_01.mp3'), soundOpts)
      .then((item) => { soundsDict['click'].isLoaded = true; });
    soundsDict['back'].audio.loadAsync(require('../assets/sounds/ButtonPress_Simple_02.mp3'), soundOpts)
      .then((item) => { soundsDict['back'].isLoaded = true; });
}
```

I tried loading only 18 sounds - this results in all 18 sounds being loaded. **However, if we load our 19th sound, ALL sounds do not load.** This causes loadAsync() to silently fail for us, which is why I figured it's related to this thread. This leads to a hang for 5-90 seconds depending on device specs.

Our workaround (which is probably a better practice anyway), is to use unloadAsync() after leaving certain screens, and use loadAsync() just in time as needed. We are aiming to only have a maximum of 12 loaded sounds concurrently to avoid the loadAsync() limit.

TL;DR: There seems to be a limit of 18 loaded sounds at once via loadAsync(). Going above this causes all sounds to fail to load. This is not a true solution because there is still an undocumented limit somewhere. Workaround is to loadAsync() when needed only and unloadAsync() when done with a sound.

**Comment by mfkrause** [comment]
> When we try to load all 26 sounds at once, we see that 0 are in fact loaded. [@mfkrause](https://github.com/mfkrause) since you mentioned preloading many sounds at once, I wonder if this is what you ran into as well.
> 
> [...]
> 
> TL;DR: There seems to be a limit of 18 loaded sounds at once via loadAsync(). Going above this causes all sounds to fail to load. This is not a true solution because there is still an undocumented limit somewhere. Workaround is to loadAsync() when needed only and unloadAsync() when done with a sound.

That's interesting. We actually do load exactly 19 sounds on mount. If that was the cause however, the problem should also be reproducible in the new `expo-audio` library, no?

I don't really know if our specific use case would allow unloading audios since we actually depend on them all on one screen and can't reliably forecast which of the audio files is needed next. I fear loading them right before they're needed could introduce too much latency. But I feel like this shouldn't be a hard limit -- even though I don't have a concrete example right now, I'd think that our app isn't the only app depending on several loaded sound files, especially when thinking of games.

Would be interesting to have some official documentation on this. I'll look into this when I have some time next week.

**Comment by singlecolor** [comment]
> That's interesting. We actually do load exactly 19 sounds on mount. If that was the cause however, the problem should also be reproducible in the new `expo-audio` library, no?

We didn't try the `expo-audio` library yet so I'm not certain. At a glance their docs seem to remove loadAsync() from the implementation, so my only guess is that loading must be handled differently by the hook-based player.

> I don't really know if our specific use case would allow unloading audios since we actually depend on them all on one screen and can't reliably forecast which of the audio files is needed next. I fear loading them right before they're needed could introduce too much latency. But I feel like this shouldn't be a hard limit -- even though I don't have a concrete example right now, I'd think that our app isn't the only app depending on several loaded sound files, especially when thinking of games.

We are in fact working on a game and definitely, we're worried about latency from the just in time calls. I can report how this is going with testers on that front later on. All iOS 18 testers just haven't even had sound yet.

> Would be interesting to have some official documentation on this. I'll look into this when I have some time next week.

Absolutely, this at least would influence a different implementation
